Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp

```diff
@@ -1457,18 +1470,28 @@ static void inverse_dct(JPEGLoadingContext const& context, Vector<Macroblock>& m
     }
 
     // F.2.1.5 - Inverse DCT (IDCT)
-
+    auto const level_shift = 1 << (context.frame.precision - 1);
+    auto const max_value = (1 << context.frame.precision) - 1;
     for (u32 vcursor = 0; vcursor < context.mblock_meta.vcount; vcursor += context.vsample_factor) {
         for (u32 hcursor = 0; hcursor < context.mblock_meta.hcount; hcursor += context.hsample_factor) {
             for (u8 vfactor_i = 0; vfactor_i < context.vsample_factor; ++vfactor_i) {
                 for (u8 hfactor_i = 0; hfactor_i < context.hsample_factor; ++hfactor_i) {
                     u32 mb_index = (vcursor + vfactor_i) * context.mblock_meta.hpadded_count + (hcursor + hfactor_i);
                     for (u8 i = 0; i < 8; ++i) {
                         for (u8 j = 0; j < 8; ++j) {
-                            macroblocks[mb_index].r[i * 8 + j] = clamp(macroblocks[mb_index].r[i * 8 + j] + 128, 0, 255);
-                            macroblocks[mb_index].g[i * 8 + j] = clamp(macroblocks[mb_index].g[i * 8 + j] + 128, 0, 255);
-                            macroblocks[mb_index].b[i * 8 + j] = clamp(macroblocks[mb_index].b[i * 8 + j] + 128, 0, 255);
-                            macroblocks[mb_index].k[i * 8 + j] = clamp(macroblocks[mb_index].k[i * 8 + j] + 128, 0, 255);
+
+                            // FIXME: This just truncate all coefficients, it's an easy way to support (read hack)
+                            //        12 bits JPEGs without rewriting all color transformations.
+                            auto const clamp_to_8_bits = [&](u16 color) -> u8 {
+                                if (context.frame.precision == 8)
+                                    return static_cast<u8>(color);
+                                return static_cast<u8>(255 * color / 4096);
```

Comment:
How did you come up with this line? From a distance it feels off.

The naive thing to do is `color >> 4`. That's uniform (there's 16 buckets in the high-res space mapped to 1 in the low-res space).

I can also see the argument that you'd want to multiply by 255 / 4095, since you want a linear map that maps 0 to 0 and 4095 (the max 12-bit value) to 255 (the max 8-bit value) (this mathy derivation ignores narrowing effects though). This would be for (morally) calling `round()` on the result, and we'd map just the first 8 values to 0, then 16 to 1, ..., and only the last 8 to 1.

But the `>> 4` evenly distributes input ranges to output ranges, so that's probably the thing to go with?

Reply:
My brain is more on the mathy side of things I guess.

But yes, `color >> 4` is probably better as it:

1. Avoids narrowing effects
2. Avoids off by one issue 4096 vs 4095 :eyes:

Let's use that instead